sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -65,6 +65,59 @@ abstract class StringRegexExpression extends BinaryExpression
   override def sql: String = s"${left.sql} ${prettyName.toUpperCase(Locale.ROOT)} ${right.sql}"
 }
 
+abstract class StringRegexV2Expression extends TernaryExpression
```

Comment:
We need this abstract class?  I think we could make the fix more simple by just tweaking `StringUtils.escapeLikeRegex`? Anyway, the simple, the better.

Reply:
Because the `StringRegexExpression`  extends the `BinaryExpression` that is only two input parameters all allowed. So I make the `StringRegexV2Expression` extends the `TernaryExpression`.

Reply:
How about just making `StringRegexExpression` ternary?

Reply:
Because RLIKE extends `StringRegexExpression` and only need two input parameters.

Reply:
Have you tried this?
```
case class Like(
  inputExpr: Expression,
  patternExpr: Expression,
  escapeExpr: Option[String] = None) extends StringRegexExpression
```

Reply:
This way exists a problem escape character could not pass to the `nullSafeCodeGen`.
```
nullSafeCodeGen(ctx, ev, (eval1, eval2) => {
        s"""
          String $patternStr = $eval2.toString();
          Character $escapeChar = ?
          $patternClass $pattern = $patternClass.compile($escapeFunc($patternStr, $escapeChar));
          ${ev.value} = $pattern.matcher($eval1.toString()).matches();
        """
      })
```
not like
```
nullSafeCodeGen(ctx, ev, (eval1, eval2, eval3) => {
	        s"""
	          String $rightStr = $eval2.toString();
	          $patternClass $pattern = $patternClass.compile($escapeFunc($rightStr));
	          String $patternStr = $eval2.toString();
	          Character $escapeChar = $eval3.toString().charAt(0);
	          $patternClass $pattern = $patternClass.compile($escapeFunc($patternStr, $escapeChar));
	          ${ev.value} = $pattern.matcher($eval1.toString()).matches();
	        """
	      })
```
If I ignore some details, tell me please.

Reply:
like this? https://github.com/apache/spark/compare/master...maropu:SPARK-28083
```
/* 066 */           String filter_rightStr_0 = scan_value_1.toString();
/* 067 */           java.util.regex.Pattern filter_pattern_0 = java.util.regex.Pattern.compile(org.apache.spark.sql.catalyst.util.StringUtils.escapeLikeRegex(filter_rightStr_0, "\"));
```

Reply:
OK, let me have a try!

Reply:
@maropu I have made this try. It was not pass the tests of `RegexpExpressionsSuite`.
The failure info:
```
- LIKE Pattern *** FAILED ***
  Code generation of null LIKE input[0, string, true] ESCAPE \ failed:

  java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 50, Column 0: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 50, Column 0: Line break in literal not allowed

  java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 50, Column 0: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 50, Column 0: Line break in literal not allowed
```

The code generated:
```
/* 033 */   public java.lang.Object apply(java.lang.Object _i) {
/* 034 */     InternalRow i = (InternalRow) _i;
/* 035 */
/* 036 */
/* 037 */     boolean isNull_0 = true;
/* 038 */     boolean value_0 = false;
/* 039 */
/* 040 */     if (!true) {
/* 041 */       boolean isNull_2 = i.isNullAt(0);
/* 042 */       UTF8String value_2 = isNull_2 ?
/* 043 */       null : (i.getUTF8String(0));
/* 044 */       if (!isNull_2) {
/* 045 */
/* 046 */         isNull_0 = false; // resultCode could change nullability.
/* 047 */
/* 048 */         String rightStr_0 = value_2.toString();
/* 049 */         java.util.regex.Pattern pattern_0 = java.util.regex.Pattern.compile(org.apache.spark.sql.catalyst.util.StringUtils.escapeLikeRegex(rightStr_0, "\"));
/* 050 */         value_0 = pattern_0.matcher(((UTF8String)null).toString()).matches();
/* 051 */
/* 052 */
/* 053 */       }
/* 054 */
/* 055 */     }
/* 056 */     isNull_3 = isNull_0;
/* 057 */     value_3 = value_0;
/* 058 */
/* 059 */     // copy all the results into MutableRow
/* 060 */
/* 061 */     if (!isNull_3) {
/* 062 */       mutableRow.setBoolean(0, value_3);
/* 063 */     } else {
/* 064 */       mutableRow.setNullAt(0);
/* 065 */     }
/* 066 */
/* 067 */     return mutableRow;
/* 068 */   }
/* 069 */
/* 070 */
```

Reply:
```
 val pattern = ctx.freshName("pattern")
      val rightStr = ctx.freshName("rightStr")
      val escapeChar = escapeCharOpt.getOrElse("\\\\")
      nullSafeCodeGen(ctx, ev, (eval1, eval2) => {
        s"""
          String $rightStr = $eval2.toString();
          $patternClass $pattern = $patternClass.compile($escapeFunc($rightStr, "$escapeChar"));
          ${ev.value} = $pattern.matcher($eval1.toString()).matches();
        """
      })
```

Reply:
I changed 
`val escapeChar = escapeCharOpt.getOrElse("\\\\")`
to
`val escapeChar = escapeCharOpt.getOrElse("\\\\\\\\")`
The latter is OK.